### DISCUSSION TITLE
Perf: avoid unnecessary window computation

### DISCUSSION THREAD

**Discussion by milancio42** [duckdb/duckdb#6242] [Ideas]
When using an unordered aggregate window function together with an ordered one, duckdb calculates the window function twice even if both use the same `PARTITION BY` clause, which is not necessary.

Consider this example:
```
explain
with a as (
select 
  unnest(['a','b','c']) as id
)
, b as (
select 
  unnest(range(5)) as ord
)
, c as (
select
  a.id
, b.ord
, round(random()*100, 2) as val
from a
cross join b
)
select
  id
, ord
, val
, sum(val) over(partition by id) as sum_a
, first_value(val) over(partition by id order by ord) as first_val
from c
;

```

result:
```
...
┌─────────────┴─────────────┐                             
│           WINDOW          │                             
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                             
│   FIRST_VALUE(val) OVER   │                             
│(PARTITION BY id ORDER...  │                             
│      ASC NULLS FIRST)     │                             
└─────────────┬─────────────┘                                                          
┌─────────────┴─────────────┐                             
│           WINDOW          │                             
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                             
│sum(val) OVER (PARTITION BY│                             
│             id)           │                             
└─────────────┬─────────────┘  
...
```

The ordering for `sum` function is irrelevant, so it can be calculated together with `first_value` function. 
This is exactly how Teradata does it.

Version of db: v0.6.1 919cad22e8

Thank you for your great work.

**Comment by hawkfish** [comment]
Good point! There is [a lot of work](https://www.comp.nus.edu.sg/~chancy/vldb12-window-function.pdf) that could be done here but this is fairly low hanging fruit. I'm about to do a serious refactoring of the window function executors and I'll keep sorting subsets in mind.

**Comment by hawkfish** [comment]
Implemented.
